### PR TITLE
Setup authentication For landlord

### PR DESCRIPTION
**Pull request by pawnjester**
#### What does this PR do?
This PR is to handle everything relating to landlord authentication (login/signup)
#### Description of Task to be completed?
- Setup Login for Landlord
- Setup Signup for Landlord
#### How should this be manually tested?
- Pull from this branch
- Include your env.
- For signup, you need the 
```
email: String,
password: String,
phoneNumber: String,
name: String
```
- For signin, you need 
```
email: String,
password: String
```
#### Any background context you want to provide?
Nil
#### What are the relevant pivotal tracker stories?(if applicable)
https://www.pivotaltracker.com/story/show/171429203
https://www.pivotaltracker.com/story/show/171429202

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,4 @@
+DB_USER=postgres
```
Comment (robocopkaka) [inline .env]: Can you ignore this file

Comment (pawnjester) [reply .env]: I ignored it but it is still showing

Comment (robocopkaka) [reply .env]: Can you do `git rm --cached`  and push a commit

File: controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import jwt from 'jsonwebtoken';
+import bcrypt from 'bcrypt';
+import models from '../models/index';
+import sequelize from 'sequelize';
+
+const Op = sequelize.Op;
+
+const user = models.User;
+
+export default class User {
```
Comment (robocopkaka) [inline controllers/userController.js]: Think this should be `Landlord` and not `User`

File: models/user.js [hunk 3]
```diff
@@ -1,10 +1,65 @@
...
+    // Instance method to prevent password from
+  // being sent to client.
+  User.prototype.toJSON = function toJSON() {
+    const values = Object.assign({}, this.get());
+
+    delete values.password;
+    return values;
+  };
+
+  // User.associate = models => {
```
Comment (robocopkaka) [inline models/user.js]: You can remove this if it's not necessary

File: controllers/landlordController.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        const newUser = await user.create({
+          email,
+          password,
+          name,
+          phoneNumber
+        })
+        const token = newUser.generateAuthToken();
+        const result = res.status(201).json({
+          statusCode: 201,
+          newUser,
```
Comment (robocopkaka) [inline controllers/landlordController.js]: Could this be changed to `user`

File: controllers/landlordController.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+          statusCode: 401,
+          error: 'Invalid Credentials'
+        })
+      }
+      const token = userFound.generateAuthToken();
+      return res.status(200).json({
+        user: 
+          userFound
+        ,
+        message: `Welcome back ${userFound.name}`,
```
Comment (robocopkaka) [inline controllers/landlordController.js:77]: Not sure if there's a need for the welcome message here but guess it's a personal preference for me so opening this to discussion. cc: @Anie96

File: middleware/validation.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { validateEmail } from './validateEmail'
+
+const validation = {
+  checkUserInput(req, res, next) {
+    let { email, password, name, phoneNumber } = req.body;
```
Comment (robocopkaka) [inline middleware/validation.js:5]: #TODO might help to add a different validation for when a field is passed that isn't among this four. Can be done later

File: tests/landlord.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+chai.should();
+const { expect } = chai;
+chai.use(chaiHttp);
+
+describe('/signup', () => {
+    before((done) => {
+
+        done()
+    });
+    after(() => db.User.destroy({
```
Comment (robocopkaka) [inline tests/landlord.js]: Any reason for this since we're already undoing migrations  before tests are run?

File: tests/landlord.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            email: 'testlogin@gmail.com',
+            password: 'testpassword',
+            name: 'testerlogin',
+            phoneNumber: '0888888888'
+        }).end((res) => {
+            done()
+    
+        })
+    });
+    after(() => db.User.destroy({
```
Comment (robocopkaka) [inline tests/landlord.js]: Same as above

File: tests/landlord.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            email: 'testlogin@gmail.com',
+            password: 'testpasswoddrd',
+        })
+        .end((error, res) => {
+            res.should.have.status(401);
+            expect(res.body.error).to.be.eq('Invalid Credentials')
+            done()
+        })
+    });
+})
```
Comment (robocopkaka) [inline tests/landlord.js]: Add newline here

**Comment by robocopkaka** [comment]
Can you update the README to show that we need to add `SECRET_KEY` to our `.env` files with a random value and also update the PR description with test objects for both endpoints to help when we pull down the branch to test

**Review by robocopkaka — APPROVED** [review]
Approved. Thanks Emmanuel

File: controllers/landlordController.js (hunk 5 above)
Comment (fire-cracker) [reply controllers/landlordController.js:77]: I am okay with the welcome message.  It's not bad idea to let your backend handle message format.

File: controllers/landlordController.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        token
+      })
+    } catch (error) {
+      return res.status(500).json({
+        statusCode: 500,
+        error: 'Error signing in'
```
Comment (fire-cracker) [inline controllers/landlordController.js]: It's great having an error message. I do think our 500 status message should be generic though. Such as "Server error"

File: middleware/validation.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    } else if (!password.trim()) {
+      return res.status(400).json({
+        "message": "The password are required.",
+        "field": "password",
+        "status": "400"
+      });
+    }
+    next()
+  },
+}
```
Comment (fire-cracker) [inline middleware/validation.js:40]: Validating for inputs is great. This can be abstracted to prevent repetition.

File: tests/landlord.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            password: 'testpassword',
+            name: 'tester',
+            phoneNumber: '0888888888'
+        })
+        .end((error, res) => {
+            res.should.have.status(201);
+            expect(res.body.user).to.have.property('email');
+            expect(res.body.user).to.have.property('name');
+            expect(res.body.user).to.have.property('phoneNumber');
+            expect(res.body.user).to.not.have.property('password');
```
Comment (fire-cracker) [inline tests/landlord.js]: Line 21  to 25 can be summarised into "expect(res.body.user).to.be.an.instanceof(Object).that.includes.all.keys('email', 'name', 'phoneNumber', 'password');"

**Review by fire-cracker — CHANGES_REQUESTED** [review]
Kindly squash commits

**Pull request merged**